Update hustings import management command for 2021

michaeljcollinsuk opened the pull request:
- Import hustings via urls by default
- Tried to keep backwards compatible so old hustings could be imported (is this necessary?)
- Optionally import from file

michaeljcollinsuk commented:
Some screenshots of updated wording of those that we have currently. Below are from the ballot view e.g. [Mayor of Bristol](https://whocanivotefor.co.uk/elections/mayor.bristol.2021-05-06/)

<img width="1552" alt="Screenshot 2021-03-23 at 18 07 32" src="https://user-images.githubusercontent.com/15347726/112197899-796ee280-8c04-11eb-8262-000b969834c7.png">

In the above you can see how past info would be displayed where it says "See video or other info from this event" (as someone seems to have entered this info already incorrectly on the [spreadsheet](https://docs.google.com/spreadsheets/d/1Krf0icsp-bExEKBpnQi_WZ6pcsT10MJ3p1NDBquf_ik/edit#gid=908428564)

<img width="1552" alt="Screenshot 2021-03-23 at 18 07 44" src="https://user-images.githubusercontent.com/15347726/112197915-7d9b0000-8c04-11eb-9b5c-7f3f94e3427d.png">
<img width="1552" alt="Screenshot 2021-03-23 at 18 14 54" src="https://user-images.githubusercontent.com/15347726/112197930-825fb400-8c04-11eb-95b4-58dea8c14de7.png">

And from a candidate/person page e.g. [https://whocanivotefor.co.uk/person/13545/sean-driscoll](https://whocanivotefor.co.uk/person/13545/sean-driscoll)
<img width="1552" alt="Screenshot 2021-03-23 at 18 15 58" src="https://user-images.githubusercontent.com/15347726/112198262-d79bc580-8c04-11eb-8021-6c0a99a3cb3c.png">

michaeljcollinsuk commented:
Are the last two checkbox items on #122 still wanted? I have also noticed that there is a "Cancelled Hustings" spreadsheet that has not been utilised - should it be?

symroe commented:
> Are the last two checkbox items on #122 still wanted? I have also noticed that there is a "Cancelled Hustings" spreadsheet that has not been utilised - should it be?

I think @pmk01 had some ideas around a page of all hustings? Also, we wouldn't need a page per husting as schema.org can have more than one event on a page - so the work would just be to wrap each row in the event mark up (or use `ld-json`).

I'll leave it up to Peter to decide on this though - my instinct is that not enough people would find it useful without it being limited to a ballot / area, but if Google were to pick them up then it would be amazing.

I _do_ have a sneaky feature request though: can you time box to ~1 hour take a stab at adding these events to the calendar response for a postcode? If it's too hard then don't worry about it, but it would be a lovely quick win.

Also, can you explain what happens when the hustings have past?

michaeljcollinsuk commented:
> Also, can you explain what happens when the hustings have past?

I kept the changes to a minimum so its not changed how they [display in past years](https://github.com/DemocracyClub/WhoCanIVoteFor/pull/698/files#diff-992d4fe7721687984d932a417ae02439eb7f05a8d7a5cc0e4931af6db9eaf7a0R5) which is just to add "(Past event)" before the event title.

It might be better to split these up in to "Upcoming events" and "Past events" sections though? Also there is no clean up of events, I could have this automated to delete and stop importing after x days after the ballot if its useful?

symroe commented:
My view is that this is fine for now - happy for Peter disagree, but also it's better to have something imperfect up sooner that we can change later than it is to try to finish it all now.

symroe commented on wcivf/apps/elections/templates/elections/includes/_single_ballot.html:
```diff
@@ -142,13 +142,14 @@ <h5>Citizenship</h5>
       <h3>
         <span aria-hidden="true">📅</span>
-        Election events near you
+        Husting events
```
In the past we've avoided using "hustings" as it's a word that I think most people understands. I would be nice to use language that explains what happens, and then include the jargon word alongside it.

pmk01 commented:
How about something like:

You can meet candidates and question them at online events (often known as 'hustings'). Here are some events where [name] may be appearing:

michaeljcollinsuk commented:
@symroe I've got iCal events working - as you can see I will now be attending four husting events for the Mayor of Bristol election
<img width="1552" alt="Screenshot 2021-03-24 at 13 04 39" src="https://user-images.githubusercontent.com/15347726/112315897-7f1b0580-8ca2-11eb-8d2e-7352aa1c2602.png">

I've also added ld-json snippets for each event as Peter agreed not worth doing a full hustings page currently. Refactored the hustings includes as I did this.

And have updated the wording as per @pmk01 suggestion

symroe commented on wcivf/apps/hustings/management/commands/import_hustings.py:
```diff
@@ -119,22 +143,14 @@ def handle(self, **options):
-                            hustings_counter, husting
-                        )
-                    )
+        count, _ = Husting.objects.all().delete()
```
I think it's good to do this, but this wouldn't work well if we tried to use the old method of throwing a file at this command - it would remove all hustings, add in the ones from the file, but then delete them again next time it was run with no arguments.

This might be really useful for local development, but I think it should come with a warning

michaeljcollinsuk replied:
Yeah this is true - it was one of the complications of trying to keep it backwards compatible. But I wasn't really sure when we would want to import from a physical file other than local dev now (I might be wrong). I've added a user prompt to confirm when using a file, do you think that is reasonable?

symroe commented on wcivf/apps/elections/templates/elections/includes/_hustings_list.html:
```diff
@@ -3,6 +3,6 @@
 <p>
 {% if husting.in_past %}(Past event) {% endif %}
-<strong>{{ husting.starts|naturalday:"l j F Y"|title }} {{ husting.starts|date:"H:i"}}{% if husting.ends %}&ndash;{{ husting.ends | date:"H:i"}}{% endif %}:</strong> {% if husting.url %}<a href="{{ husting.url }}">{% endif %}{{ husting.title}}{% if husting.url %}</a>{% endif %}, {{ husting.location }} {{ husting.postcode }}. {% if husting.postevent_url %}<a href="{{ husting.postevent_url }}"><i c ...
+<strong>{{ husting.starts|naturalday:"l j F Y"|title }} {{ husting.starts|date:"H:i"}}{% if husting.ends %}&ndash;{{ husting.ends | date:"H:i"}}{% endif %}:</strong> {% if husting.url %}<a href="{{ husting.url }}">{% endif %}{{ husting.title}}{% if husting.url %}</a>{% endif %}{% if husting.location %}, {{ husting.location }} {{ husting.postcode }}{% endif %}{% if husting.postevent_url %}<br><a h ...
```
I know this is how it was previously, but can we have some line breaks here, for legibility?
